server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -369,34 +385,31 @@ private void extractItem(String path, XContentParser source) throws IOException
                         hashes.add(new NameAndHash(new BytesRef(path), hash(new BytesRef(source.text()))));
                         source.nextToken();
                         break;
+                    case START_ARRAY:
+                        source.nextToken();
+                        extractArray(path, source);
+                        source.nextToken();
+                        break;
                     case VALUE_NULL:
                         source.nextToken();
                         break;
                     default:
                         throw new ParsingException(
                             source.getTokenLocation(),
-                            "Routing values must be strings but found [{}]",
+                            "Cannot extract routing path due to unexpected token [{}]",
                             source.currentToken()
                         );
                 }
             }
 
             private int buildHash(IntSupplier onEmpty) {
-                Collections.sort(hashes);
-                Iterator<NameAndHash> itr = hashes.iterator();
-                if (itr.hasNext() == false) {
+                if (hashes.isEmpty()) {
                     return onEmpty.getAsInt();
                 }
-                NameAndHash prev = itr.next();
-                int hash = hash(prev.name) ^ prev.hash;
-                while (itr.hasNext()) {
-                    NameAndHash next = itr.next();
-                    if (prev.name.equals(next.name)) {
-                        throw new IllegalArgumentException("Duplicate routing dimension for [" + next.name + "]");
-                    }
-                    int thisHash = hash(next.name) ^ next.hash;
-                    hash = 31 * hash + thisHash;
-                    prev = next;
+                Collections.sort(hashes);
```

Comment:
Do I understand correctly that array values will have the same name? If so, how do we guarantee that they won't be shuffled by sorting - default sorting is not stable?  

One option is to add a suffix to the name for array elements, e.g. `.0`, `.1` etc. This way, you don't need to touch this function?

Reply:
I solved a bit differently. But the ordering should be guaranteed now.